File: 106. Construct Binary Tree from Inorder and Postorder Traversal/buildTree.js
// Definition for a binary tree node.
class TreeNode {
  constructor(val = 0, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
 }
/**
 * @param {number[]} inorder
 * @param {number[]} postorder
 * @return {TreeNode}
 */
const buildTree = (inorder, postorder) => {
  const inToIndex = {};
  for (let i = 0; i < inorder.length; i++) {
    inToIndex[inorder[i]] = i;
  }

  const build = (inStart, inEnd, posStart, posEnd) => {
    if (inStart > inEnd) {
      return null;
    }

    const rootVal = postorder[posEnd];
    const rootInIndex = inToIndex[rootVal];
    const leftSize = rootInIndex - inStart;

    const root = new TreeNode(rootVal);
    root.left = build(inStart, rootInIndex - 1, posStart, posStart + leftSize - 1);
    root.right = build(rootInIndex + 1, inEnd, posStart + leftSize, posEnd - 1);

    return root;
  }

  return build(0, inorder.length - 1, 0, postorder.length - 1);
};


const inorder = [9, 3, 15, 20, 7];
const postorder = [9, 15, 7, 20, 3];
console.log("buildTree: ", buildTree(inorder, postorder))

File: 106. Construct Binary Tree from Inorder and Postorder Traversal/buildTree.py
from typing import List, Optional

# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right

class Solution:
  def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
    inToIndex = { num: i for i, num in enumerate(inorder) }
    
    def build(inStart, inEnd, posStart, posEnd):
      if inStart > inEnd:
        return None

      rootVal = postorder[posEnd]
      rootInIndex = inToIndex[rootVal]
      leftSize = rootInIndex - inStart
      
      root = TreeNode(rootVal)
      root.left = build(inStart, rootInIndex - 1, posStart, posStart + leftSize - 1)
      root.right = build(rootInIndex + 1, inEnd, posStart + leftSize, posEnd - 1)
      
      return root
    
    return build(0, len(inorder) - 1, 0, len(postorder) - 1)
    

inorder = [9, 3, 15, 20, 7]
postorder = [9, 15, 7, 20, 3]
solution = Solution()
print("buildTree: ", solution.buildTree(inorder, postorder))
